[PowerPC] Support local-dynamic TLS relocation on AIX

orcguru opened the pull request:
Supports TLS local-dynamic on AIX, generates below sequence of code:

```
.tc foo[TC],foo[TL]@ld # Variable offset, ld relocation specifier
.tc mh[TC],mh[TC]@ml # Module handle for the caller
lwz 3,mh[TC]\(2\) $$ For 64-bit: ld 3,mh[TC]\(2\)
bla .__tls_get_mod # Modifies r0,r3,r4,r5,r11,lr,cr0
#r3 = &TLS for module
lwz 4,foo[TC]\(2\) $$ For 64-bit: ld 4,foo[TC]\(2\)
add 5,3,4 # Compute &foo
.rename mh[TC], "\_$TLSML" # Symbol for the module handle must have the name "_$TLSML"
```

Commit:
[PowerPC] Support local-dynamic TLS relocation on AIX

bzEq commented on llvm/lib/Target/PowerPC/PPCISelLowering.h:
```diff
@@ -370,11 +370,23 @@ namespace llvm {
 
+    /// %x3 = GET_TLS_MOD _$TLSML - For the AIX local-dynamic TLS model,
+    /// produces a call to __tls_get_mod(_$TLSML\@ml).
+    GET_TLS_MOD,
```
Rename to `GET_TLS_MOD_AIX`.

orcguru replied:
Thanks! I will update in the next commit.

bzEq commented on llvm/lib/Target/PowerPC/PPCInstr64Bit.td:
```diff
@@ -1584,12 +1584,19 @@ def GETtlsldADDRPCREL : GETtlsldADDRPseudo <"#GETtlsldADDRPCREL">;
                     "GETtlsADDR64AIX",
                     [(set i64:$rD,
                       (PPCgetTlsAddr i64:$offset, i64:$handle))]>, isPPC64;
+// On AIX, the call to __tls_get_mod need one input in X3 for the module handle.
```
Is it missing `Uses = [X3]` for this pseudo instruction?

chenzheng1030 replied:
```suggestion
// On AIX, the call to .__tls_get_mod need one input in X3 for the module handle.
```

orcguru commented on llvm/test/CodeGen/PowerPC/aix-tls-gd-double.ll:
```diff
@@ -156,10 +156,11 @@ define void @storesTIInit(double %Val) #0 {
-; SMALL32-NEXT:    lwz 4, L..C5(2) # target-flags(ppc-tlsgd) @TIInit
 ; SMALL32-NEXT:    stw 0, 40(1)
-; SMALL32-NEXT:    bla .__tls_get_addr[PR]
+; SMALL32-NEXT:    lwz 6, L..C4(2) # target-flags(ppc-tlsld) @TIInit
```
These cases changed because `TargetMachine::getTLSModel()` decides TIInit model is TLSModel::LocalDynamic. Before this patch, this case is handled by the default GD model in `PPCTargetLowering::LowerGlobalTLSAddressAIX()`. Now with LD support, they changed model.

bzEq replied:
Better to post an NFC patch to explicitly specify these thread local variables to be `general-dynamic` similar to what you do in `llvm/test/CodeGen/PowerPC/aix-tls-local-dynamic.ll`, so they are not affected by this PR.

orcguru replied:
I'm afraid `generaldynamic` is not a valid explicit mode according to document: https://llvm.org/docs/LangRef.html
```
A variable may be defined as thread_local, which means that it will not be shared by threads (each thread will have a separated copy of the variable). Not all targets support thread-local variables. Optionally, a TLS model may be specified:

localdynamic

    For variables that are only used within the current shared library.
initialexec

    For variables in modules that will not be loaded dynamically.
localexec

    For variables defined in the executable and only used within it.

If no explicit model is given, the “general dynamic” model is used.
```

bzEq replied:
How about using `llc --thread-model=...`?

bzEq replied:
How about adding `-relocation-model=pic` to RUN lines?

bzEq replied:
I just notice the change here is related to an `internal` symbol, so the change is expected after local-dynamic is enabled.

orcguru commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,19 +158,55 @@ namespace {
+                .addExternalSymbol("_$TLSML[TC]", PPCII::MO_TLSLD_FLAG)
+                .addReg(IsLargeModel
+                            ? ModuleHandleHReg
+                            : Register(Subtarget.getTOCPointerRegister()));
```
I have to explicitly call `Register()` to convert MCRegister into Register type, otherwise compiler complains type mismatch due to the usage of `?:`.

Commit:
address comments

bzEq commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -846,6 +870,15 @@ void PPCAsmPrinter::emitInstruction(const MachineInstr *MI) {
+        // a TOC symbol from the symbol itself, and right now its only user is
+        // symbol "_$TLSML". Use symbol name to decide that R_TLSML is expected.
+        return MCSymbolRefExpr::VariantKind::VK_PPC_AIX_TLSML;
+      if (IsAIX)
```
Is it redundant in the context?

orcguru replied:
I have to hitchhike the `VK_PPC_AIX_TLSML` under `PPCII::MO_TLSLD_FLAG` as  PPCII TOF is already full, and due to the relevance with TLSLD, I prefer to add the check here.

bzEq commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -846,6 +870,15 @@ void PPCAsmPrinter::emitInstruction(const MachineInstr *MI) {
     if (MO.getTargetFlags() & PPCII::MO_TLSGD_FLAG)
       return MCSymbolRefExpr::VariantKind::VK_PPC_AIX_TLSGD;
+    if (MO.getTargetFlags() & PPCII::MO_TLSLD_FLAG) {
+      if (isSpecialAIXSymbolTLSML(MO, IsAIX))
```
Why not just compare name at the beginning rather than checking TargetFlag?

orcguru replied:
The symbol is specific for TLSLD, so I prefer to encapsulate the logic inside TLSLD relevant predator.

bzEq replied:
```suggestion
      if (isAIX && isTLSMLSymbolOperand(MO))
```
`bool` in function's argument list is not preferred.

orcguru replied:
Oh, I saw in https://llvm.org/docs/ProgrammersManual.html there is function which take bool as argument. I won't change unless somebody change that document.

Commit:
[NFC] address comments.

orcguru commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -660,14 +671,15 @@ void PPCAsmPrinter::EmitTlsCall(const MachineInstr *MI,
     (void)VarOffsetReg;
-    assert(MI->getOperand(2).isReg() &&
-           MI->getOperand(2).getReg() == VarOffsetReg &&
+    assert((MI->getNumExplicitOperands() < 3 ||
```
I changed the assert a bit to explicitly allow the cases for `.__tls_get_mod`

ecnelises commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -660,15 +671,17 @@ void PPCAsmPrinter::EmitTlsCall(const MachineInstr *MI,
-    // should already be in R3.
-    // For TLSGD, which currently is the only supported access model, we only
-    // need to generate an absolute branch to .__tls_get_addr.
+    // On AIX, for TLSGD the variable offset should already be in R4 and the
+    // region handle should already be in R3, need to generate an absolute
+    // branch to .__tls_get_addr. For TLSLD the module handle should already be
+    // in R3, need to generate branch to .__tls_get_mod.
```
```suggestion
    // For TLSGD, the variable offset should already be in R4 and the region
    // handle should already be in R3, generate absolute branch to .__tls_get_addr.
    // For TLSLD, the module handle should already be in R3, generate branch
    // to .__tls_get_mod.
```

ecnelises commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -660,15 +671,17 @@ void PPCAsmPrinter::EmitTlsCall(const MachineInstr *MI,
-    assert(MI->getOperand(2).isReg() &&
-           MI->getOperand(2).getReg() == VarOffsetReg &&
-           "GETtls[ld]ADDR[32] must read GPR4");
+    assert(MI->getOpcode() == PPC::GETtlsMOD32AIX ||
+           MI->getOpcode() == PPC::GETtlsMOD64AIX ||
+           (MI->getOperand(2).isReg() &&
+            MI->getOperand(2).getReg() == VarOffsetReg) &&
+               "GETtls[ld]ADDR[32] must read GPR4");
```
```suggestion
    assert((MI->getOpcode() == PPC::GETtlsMOD32AIX ||
            MI->getOpcode() == PPC::GETtlsMOD64AIX ||
            (MI->getOperand(2).isReg() &&
             MI->getOperand(2).getReg() == VarOffsetReg)) &&
           "GETtls[ld]ADDR[32] must read GPR4");
```

`&&` has higher precedence than `||`.

orcguru replied:
To be honest, I looked up the C operator precedence. My intention here is to skip the assert check for those two opcodes. On the other hand, I think to be a good programmer for LLVM, maybe I should follow your suggestion since I didn't find similar usage like this...

ecnelises commented on llvm/lib/Target/PowerPC/PPCISelLowering.cpp:
```diff
@@ -3412,13 +3416,25 @@ SDValue PPCTargetLowering::LowerGlobalTLSAddressAIX(SDValue Op,
+    return DAG.getNode(PPCISD::TLSLD_AIX, dl, PtrVT, VariableOffset);
+  }
+
+  // The Local-Exec, Initial-Exec, Local-Dynamic, and General-Dynamic TLS models
+  // are currently supported access models. If Local- or Initial-exec or
```
> _The Local-Exec, Initial-Exec, Local-Dynamic, and General-Dynamic TLS models are currently supported access models._

The four are all models. This sentence can be removed.

Commit:
[NFC] address comments

orcguru commented:
@bzEq @ecnelises If there is no major concern, can we land this first before the weekend? Thank you!

chenzheng1030 commented on llvm/test/CodeGen/PowerPC/aix-tls-gd-double.ll:
```diff
@@ -156,10 +156,11 @@ define void @storesTIInit(double %Val) #0 {
 ; SMALL32:       # %bb.0: # %entry
```
These cases(with `-gd-` in their names) are seems trying to test global-dynamic mode, so maybe it makes more sense we keep the original test points and add new runtime for the local-dynamic mode.

bzEq replied:
The change is subtle, since `TIInit` is `internal`, all changes in this file is about `TIInit`.

orcguru replied:
Looks like there are some overlapping test coverage in these TLS related cases. Maybe after enable local-dynamic, we could go through a process to do some cleanup or move some cases to their own file.

chenzheng1030 commented on llvm/test/CodeGen/PowerPC/aix-tls-xcoff-reloc-large.ll:
```diff
@@ -32,7 +33,7 @@ entry:
 ; RELOC-NEXT:   Relocation {
 ; RELOC-NEXT:     Virtual Address: 0x16
-; RELOC-NEXT:     Symbol: .TIInit (17)
+; RELOC-NEXT:     Symbol: TIInit (19)
```
I know this is not this patch's scope. But it may make more sense to use a RE to handle the symbol index. We were trying to fix this, but I guess we didn't change them all.

chenzheng1030 reviewed (commented):
Some comments may need follow up, so I posted them first. I will continue to review the code generation part.

We need a front end patch to eliminate the error report in `clang/lib/Frontend/CompilerInvocation.cpp` and `clang/lib/Sema/SemaDeclAttr.cpp`, right?

chenzheng1030 commented on llvm/test/CodeGen/PowerPC/aix-tls-xcoff-reloc-large.ll:
```diff
@@ -240,15 +241,35 @@ entry:
 ; SYM-NEXT:   Symbol {
 ; SYM-NEXT:     Index: 3
-; SYM-NEXT:     Name:
+; SYM-NEXT:     Name: .__tls_get_addr
```
The comment `I know this is not this patch's scope. But it may make more sense to use a RE to handle the symbol index. We were trying to fix this, but I guess we didn't change them all.` is more appropriate here.

Commit:
[NFC] address comment

orcguru commented:
> Some comments may need follow up, so I posted them first. I will continue to review the code generation part.
> 
> We need a front end patch to eliminate the error report in `clang/lib/Frontend/CompilerInvocation.cpp` and `clang/lib/Sema/SemaDeclAttr.cpp`, right?

I tried to parallelize the review effort, but github didn't show the delta for me: https://github.com/llvm/llvm-project/pull/66972

bzEq commented on llvm/test/CodeGen/PowerPC/aix-tls-gd-double.ll:
```diff
@@ -156,10 +156,11 @@ define void @storesTIInit(double %Val) #0 {
-; SMALL32-NEXT:    lwz 4, L..C5(2) # target-flags(ppc-tlsgd) @TIInit
 ; SMALL32-NEXT:    stw 0, 40(1)
-; SMALL32-NEXT:    bla .__tls_get_addr[PR]
+; SMALL32-NEXT:    lwz 6, L..C4(2) # target-flags(ppc-tlsld) @TIInit
```
Can we use `r4` here? That should align with the assembler manual and save register use.

orcguru replied:
Updated a version to address the issue, will continue improve this point.

orcguru replied:
New approach will invalidate [d400de0](https://github.com/llvm/llvm-project/pull/66316/commits/d400de0cd3ffbb80794961a3443a191760b32aa1), please ignore that.

orcguru replied:
Now for the small-model, we did reuse r4. However for the large-model, since machine-scheduler will decide to move the ADDIS before the .__tls_get_mod call, I did not try to move ADDIS in PPCTLSDynamicCall.cpp

bzEq commented on llvm/lib/Target/PowerPC/PPCISelLowering.cpp:
```diff
@@ -3412,13 +3416,23 @@ SDValue PPCTargetLowering::LowerGlobalTLSAddressAIX(SDValue Op,
+    // PPCTLSDynamicCall.
+    SDValue VariableOffsetTGA =
+        DAG.getTargetGlobalAddress(GV, dl, PtrVT, 0, PPCII::MO_TLSLD_FLAG);
+    SDValue VariableOffset = getTOCEntry(DAG, dl, VariableOffsetTGA);
```
Is it missing TOC entry for the module handle?

chenzheng1030 replied:
+1. We can add another flag `PPCII::MO_TLSLDM_FLAG`(Maybe there is not enough bit in `PPCII`, we can use a distinct value, or extend the MO_ACCESS_MASK to free more bits.).

With this way, seems we don't need the "strange" function `isSpecialAIXSymbolTLSML()` in the PPCAsmPrinter file.

amy-kwan replied:
I agree, I was thinking we need the TOC entry for the module handle.

orcguru replied:
There is a blocker which limits FLAG bit size to be 12:

```
llvm/include/llvm/CodeGen/MachineOperand.h
...
  unsigned SubReg_TargetFlags : 12;
...
  void setTargetFlags(unsigned F) {
    assert(!isReg() && "Register operands can't have target flags");
    SubReg_TargetFlags = F;
    assert(SubReg_TargetFlags == F && "Target flags out of range");
  } 
```

Currently PPCII::TOF is already full. I'm afraid this will be a FIXME...

orcguru replied:
I still didn't figure out how to remove the strange function.

The way I tried was getOrCreate ExternalLinkage GV named "_$TLSML", and call getTOCEntry on that GV. However there will be a symbol table entry for the symbol, but we don't want that.

For asm mode, I can add hack in MCAsmStreamer::emitXCOFFSymbolLinkageWithVisibility to do not emit the ".extern _Renamed..5f24__TLSML; rename _Renamed..5f24__TLSML ......".

For obj mode, things are more complex, and I'm not able to cleanly remove the symbol table entry without by using hack approach.

bzEq replied:
I don't think the module handle has external linkage, `_$TLSML` is renamed from the module handle defined in the module. If we need a `GlobalValue` to create a TOC entry for the module handle, `PreISelIntrinsicLowering` would be a candidate pass to realize it.

orcguru replied:
So I did have another version which can essentially remove the assert check with some side effects:

(1) It did generated extra external symbol reference, which looks like no harm with simple HelloWorld. (Notice: xlC does not generate this symbol reference.)
```
        .extern _Renamed..5f24__TLSML[UA]
        .rename _Renamed..5f24__TLSML[UA],"_$TLSML"
```
(2) Inside `XCOFFObjectWriter::recordRelocation()` logic I have to apply a hack to get the correct symbol, because now there are two symbols: one is `_Renamed..5f24__TLSML[UA]`, the other is `_Renamed..5f24__TLSML[TC]`. Without hack it will choose the [UA] one which will cause linker failure.

Attached the draft patch: [ReplaceAssertHack.diff.txt](https://github.com/llvm/llvm-project/files/12723368/ReplaceAssertHack.diff.txt)

I think current version has the advantage that it does not generate any extra external symbol reference or any extra relocation entry, although it looks ugly with regard to those asserts.

Is it possible that we fix the hack implementation with some future version (I will open an issue and work on it afterwards)? How about let's move on with current approach? 

@amy-kwan @stephenpeckham @bzEq @chenzheng1030 Appreciate your comments. Thank you!

chenzheng1030 replied:
4b932d84f48e0f3f42c769a5ca7ce6623ab62f2e is committed to redesign the target flags of machine operand on PPC. Now we should be able to add new flags.

bzEq commented:
> I tried to parallelize the review effort, but github didn't show the delta for me: https://github.com/llvm/llvm-project/pull/66972

Let's include FE change in this PR.

chenzheng1030 reviewed (commented):
For below case:
```
__thread extern  int x = 42;
__thread extern int y = 42;
int main(void) {
   return x + y;
}
```

For now, we get:
```
.main:
# %bb.0:                                # %entry
        mflr 0
        stwu 1, -32(1)
        stw 0, 40(1)
        lwz 6, L..C0(2)                         # target-flags(ppc-tlsld) @x
        lwz 7, L..C1(2)                         # target-flags(ppc-tlsld) @y
        lwz 3, L..C2(2)                         # target-flags(ppc-tlsld) &"_$TLSML[TC]"
        bla .__tls_get_mod[PR]
        add 3, 3, 6
        lwz 6, 0(3)
        lwz 3, L..C2(2)                         # target-flags(ppc-tlsld) &"_$TLSML[TC]"
        bla .__tls_get_mod[PR]
        add 3, 3, 7
        lwz 3, 0(3)
        add 3, 3, 6
        addi 1, 1, 32
        lwz 0, 8(1)
        mtlr 0
        blr
```

I think we should be able to save one `__tls_get_mod()`?

If this is not easy to optimize now, I'm OK we do this optimization later.

stephenpeckham commented:
The primary advantage of the local-dynamic access method is that you only make a single function call to __tls_get_mod() and use the returned pointer to get at all TLS variables.  This implementation makes a function call for every variable.  The only thing you've gained is a reduction in the number of TOC symbols.

amy-kwan commented on llvm/lib/Target/PowerPC/PPCInstr64Bit.td:
```diff
@@ -1584,12 +1584,19 @@ def GETtlsldADDRPCREL : GETtlsldADDRPseudo <"#GETtlsldADDRPCREL">;
                     "GETtlsADDR64AIX",
                     [(set i64:$rD,
                       (PPCgetTlsAddr i64:$offset, i64:$handle))]>, isPPC64;
+// On AIX, the call to .__tls_get_mod need one input in X3 for the module handle.
```
```suggestion
// On AIX, the call to .__tls_get_mod needs one input in X3 for the module handle.
```

amy-kwan reviewed (commented):
Some small initial comments first.

Commit:
Attempt to address comment: use r4 for LoadOffsetToc

However looks like machine-scheduler is interfering, and still schedule
LoadOffsetToc ahead of the .__tls_get_mod call.

orcguru commented on llvm/test/CodeGen/PowerPC/aix-tls-gd-double.ll:
```diff
@@ -172,10 +173,12 @@ define void @storesTIInit(double %Val) #0 {
-; LARGE32-NEXT:    lwz 4, L..C5@l(4)
-; LARGE32-NEXT:    bla .__tls_get_addr[PR]
+; LARGE32-NEXT:    addis 4, L..C5@u(2)
+; LARGE32-NEXT:    lwz 6, L..C5@l(4)
```
# *** IR Dump After PowerPC TLS Dynamic Call Fixup (ppc-tls-dynamic-call) ***:
# Machine code for function storesTIInit: IsSSA, TracksLiveness
Function Live Ins: $f1 in %0

bb.0.entry:
  liveins: $f1
  %0:f8rc = COPY $f1
  %4:g8rc_and_g8rc_nox0 = ADDIStocHA8 $x2, target-flags(ppc-tlsld) &"_$TLSML[TC]"
  %5:g8rc_and_g8rc_nox0 = LDtocL target-flags(ppc-tlsld) &"_$TLSML[TC]", %4:g8rc_and_g8rc_nox0
  $x3 = COPY %5:g8rc_and_g8rc_nox0
  $x3 = GETtlsMOD64AIX $x3, implicit-def $x0, implicit-def $x4, implicit-def $x5, implicit-def $x11, implicit-def $lr8, implicit-def $cr0
  %1:g8rc_and_g8rc_nox0 = ADDIStocHA8 $x2, target-flags(ppc-tlsld) @TIInit
  %2:g8rc = LDtocL target-flags(ppc-tlsld) @TIInit, %1:g8rc_and_g8rc_nox0 :: (load (s64) from got)
  ADJCALLSTACKDOWN 0, 0, implicit-def $r1, implicit $r1
  $x4 = COPY %2:g8rc
  $x3 = ADD8 $x3, $x4
  ADJCALLSTACKUP 0, 0, implicit-def $r1, implicit $r1
  %3:g8rc_and_g8rc_nox0 = COPY $x3
  STFD %0:f8rc, 0, %3:g8rc_and_g8rc_nox0 :: (store (s64) into @TIInit)
  BLR8 implicit $lr8, implicit $rm

# End machine code for function storesTIInit.


# *** IR Dump After Machine Instruction Scheduler (machine-scheduler) ***:
# Machine code for function storesTIInit: NoPHIs, TracksLiveness, TiedOpsRewritten
Function Live Ins: $f1 in %0

0B      bb.0.entry:
          liveins: $f1
16B       %0:f8rc = COPY $f1
32B       %4:g8rc_and_g8rc_nox0 = ADDIStocHA8 $x2, target-flags(ppc-tlsld) &"_$TLSML[TC]"
48B       %5:g8rc_and_g8rc_nox0 = LDtocL target-flags(ppc-tlsld) &"_$TLSML[TC]", %4:g8rc_and_g8rc_nox0, implicit $x2
96B       %1:g8rc_and_g8rc_nox0 = ADDIStocHA8 $x2, target-flags(ppc-tlsld) @TIInit
100B      $x3 = COPY %5:g8rc_and_g8rc_nox0
104B      $x3 = GETtlsMOD64AIX $x3, implicit-def dead $x0, implicit-def dead $x4, implicit-def dead $x5, implicit-def dead $x11, implicit-def $lr8, implicit-def dead $cr0
112B      %2:g8rc = LDtocL target-flags(ppc-tlsld) @TIInit, %1:g8rc_and_g8rc_nox0, implicit $x2 :: (load (s64) from got)
128B      ADJCALLSTACKDOWN 0, 0, implicit-def $r1, implicit $r1
144B      $x4 = COPY %2:g8rc
160B      $x3 = ADD8 $x3, $x4
176B      ADJCALLSTACKUP 0, 0, implicit-def $r1, implicit $r1
192B      %3:g8rc_and_g8rc_nox0 = COPY $x3
208B      STFD %0:f8rc, 0, %3:g8rc_and_g8rc_nox0 :: (store (s64) into @TIInit)
224B      BLR8 implicit $lr8, implicit $rm

# End machine code for function storesTIInit.

orcguru commented on llvm/test/CodeGen/PowerPC/aix-tls-gd-int.ll:
```diff
@@ -181,10 +182,12 @@ define void @storesTIUninit(i32 %Val) #0 {
-; LARGE32-NEXT:    bla .__tls_get_addr[PR]
+; LARGE32-NEXT:    addis 4, L..C5@u(2)
+; LARGE32-NEXT:    lwz 7, L..C5@l(4)
+; LARGE32-NEXT:    bla .__tls_get_mod[PR]
```
```
# *** IR Dump After PowerPC TLS Dynamic Call Fixup (ppc-tls-dynamic-call) ***:
# Machine code for function storesTIUninit: IsSSA, TracksLiveness
Function Live Ins: $r3 in %0

bb.0.entry:
  liveins: $r3
  %0:gprc = COPY $r3
  %4:gprc_and_gprc_nor0 = ADDIStocHA $r2, target-flags(ppc-tlsld) &"_$TLSML[TC]"
  %5:gprc_and_gprc_nor0 = LWZtocL target-flags(ppc-tlsld) &"_$TLSML[TC]", %4:gprc_and_gprc_nor0
  $r3 = COPY %5:gprc_and_gprc_nor0
  $r3 = GETtlsMOD32AIX $r3, implicit-def $r0, implicit-def $r4, implicit-def $r5, implicit-def $r11, implicit-def $lr, implicit-def $cr0
  %1:gprc_and_gprc_nor0 = ADDIStocHA $r2, target-flags(ppc-tlsld) @TIUninit
  %2:gprc = LWZtocL target-flags(ppc-tlsld) @TIUninit, %1:gprc_and_gprc_nor0 :: (load (s32) from got)
  ADJCALLSTACKDOWN 0, 0, implicit-def $r1, implicit $r1
  $r4 = COPY %2:gprc
  $r3 = ADD4 $r3, $r4
  ADJCALLSTACKUP 0, 0, implicit-def $r1, implicit $r1
  %3:gprc_and_gprc_nor0 = COPY $r3
  STW %0:gprc, 0, %3:gprc_and_gprc_nor0 :: (store (s32) into @TIUninit)
  BLR implicit $lr, implicit $rm

# End machine code for function storesTIUninit.
```

```
# *** IR Dump After Machine Instruction Scheduler (machine-scheduler) ***:
# Machine code for function storesTIUninit: NoPHIs, TracksLiveness, TiedOpsRewritten
Function Live Ins: $r3 in %0

0B      bb.0.entry:
          liveins: $r3
16B       %0:gprc = COPY $r3
32B       %4:gprc_and_gprc_nor0 = ADDIStocHA $r2, target-flags(ppc-tlsld) &"_$TLSML[TC]"
48B       %5:gprc_and_gprc_nor0 = LWZtocL target-flags(ppc-tlsld) &"_$TLSML[TC]", %4:gprc_and_gprc_nor0, implicit $r2
56B       %1:gprc_and_gprc_nor0 = ADDIStocHA $r2, target-flags(ppc-tlsld) @TIUninit
64B       $r3 = COPY %5:gprc_and_gprc_nor0
72B       %2:gprc = LWZtocL target-flags(ppc-tlsld) @TIUninit, %1:gprc_and_gprc_nor0, implicit $r2 :: (load (s32) from got)
80B       $r3 = GETtlsMOD32AIX $r3, implicit-def dead $r0, implicit-def dead $r4, implicit-def dead $r5, implicit-def dead $r11, implicit-def $lr, implicit-def dead $cr0
128B      ADJCALLSTACKDOWN 0, 0, implicit-def $r1, implicit $r1
144B      $r4 = COPY %2:gprc
160B      $r3 = ADD4 $r3, killed $r4
176B      ADJCALLSTACKUP 0, 0, implicit-def $r1, implicit $r1
192B      %3:gprc_and_gprc_nor0 = COPY $r3
208B      STW %0:gprc, 0, %3:gprc_and_gprc_nor0 :: (store (s32) into @TIUninit)
224B      BLR implicit $lr, implicit $rm

# End machine code for function storesTIUninit.
```

Commit:
Remove TLS local-dynamic mode guards for AIX.

orcguru commented:
> > I tried to parallelize the review effort, but github didn't show the delta for me: #66972
> 
> Let's include FE change in this PR.

Thanks! Added clang change.

orcguru commented:
> The primary advantage of the local-dynamic access method is that you only make a single function call to __tls_get_mod() and use the returned pointer to get at all TLS variables. This implementation makes a function call for every variable. The only thing you've gained is a reduction in the number of TOC symbols.

The issue with duplicated calls to .__tls_get_mod is due to those pseudo TLSLDAIX nodes depending on different GVs, and CSE is not able to do dedup. I probably have to create a pseudo GV named like "_$TLSML", and use that GV for all TLSLDAIX pseudo nodes.

Commit:
Fixed issues raised by comments and incorporated suggested changes.
(1) Removed TLSLDAIX argument so that duplicated nodes can be eliminated.
(2) Try to reuse clobber r4 by moving TLSLDAIX ahead of LoadOffsetToc node.
(3) Add FIXME comments.
(4) Add test case to show duplicated .__tls_get_mod can be eliminated.

Below two cases are not updated yet due to environment issue. I will fix
those in next update.
Failed Tests:
  LLVM :: CodeGen/PowerPC/aix-tls-xcoff-reloc-large.ll
  LLVM :: CodeGen/PowerPC/aix-tls-xcoff-reloc.ll

Commit:
(1) Use GPR3 directly in LoadModuleHandle.
(2) Update comments.

orcguru commented on llvm/lib/Target/PowerPC/PPCInstr64Bit.td:
```diff
@@ -1622,6 +1629,10 @@ def TLSGDAIX8 :
                        (PPCTlsgdAIX i64:$offset, i64:$handle))]>;
+// This pseudo is expanded to one copy to put the module handle in R3, then call
+// GETtlsMOD64AIX, and then add variable offset to the output from the call.
+def TLSLDAIX8 : PPCEmitTimePseudo<(outs g8rc:$rD), (ins g8rc:$handle),
```
Thank you. I have changed the implementation of the pseudo node: it loads module-handle into R/X3 and then calls .__tls_get_mod. (ADD is implemented by standalone ISD::ADD, so not in the scope of this pseudo node)

Now I changed the comment to:
// This pseudo is expanded to load module-handle in X3, and the call to GETtlsMOD64AIX.

orcguru commented on llvm/lib/Target/PowerPC/MCTargetDesc/PPCMCTargetDesc.cpp:
```diff
@@ -231,12 +231,15 @@ class PPCTargetAsmStreamer : public PPCTargetStreamer {
               ->getQualNameSymbol();
-      // On AIX, we have a region handle (symbol@m) and the variable offset
-      // (symbol@{gd|ie|le}) for TLS variables, depending on the TLS model.
+      // On AIX, we have a region handle (symbol@m), module handle
```
Thank you! I put this description into the comment.

orcguru commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -1362,6 +1402,8 @@ void PPCAsmPrinter::emitInstruction(const MachineInstr *MI) {
   case PPC::GETtlsADDRPCREL:
   case PPC::GETtlsADDR32AIX:
   case PPC::GETtlsADDR64AIX:
+  case PPC::GETtlsMOD32AIX:
```
Added following description:
    // Transform: %r3 = GETtlsMODNNAIX %r3 (for NN == 32/64).
    // Into: BLA .__tls_get_mod()
    // Input parameter is a module handle (__TLSML[TC]@ml) for all variables.

orcguru commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -2826,6 +2877,8 @@ void PPCAIXAsmPrinter::emitInstruction(const MachineInstr *MI) {
+  case PPC::GETtlsMOD64AIX:
   case PPC::GETtlsTpointer32AIX:
   case PPC::GETtlsADDR64AIX:
   case PPC::GETtlsADDR32AIX: {
```
Added comments:
    // A reference to .__tls_get_mod is unknown to the assembler so we need to
    // emit an external symbol reference.

orcguru commented on llvm/lib/Target/PowerPC/PPCISelLowering.cpp:
```diff
@@ -1771,9 +1771,13 @@ const char *PPCTargetLowering::getTargetNodeName(unsigned Opcode) const {
   case PPCISD::ADDIS_TLSGD_HA:  return "PPCISD::ADDIS_TLSGD_HA";
   case PPCISD::ADDI_TLSGD_L:    return "PPCISD::ADDI_TLSGD_L";
   case PPCISD::GET_TLS_ADDR:    return "PPCISD::GET_TLS_ADDR";
+  case PPCISD::GET_TLS_MOD_AIX:
```
Sure. I followed your advice. Thank you!

orcguru commented on llvm/lib/Target/PowerPC/PPCISelLowering.cpp:
```diff
@@ -3412,13 +3416,23 @@ SDValue PPCTargetLowering::LowerGlobalTLSAddressAIX(SDValue Op,
-  // handle is created with the MO_TLSGDM_FLAG flag and the global address
-  // for the TOC entry of the variable offset is created with MO_TLSGD_FLAG.
+  if (Model == TLSModel::LocalDynamic) {
+    // For local-dynamic on AIX, we need to generate two TOC entries, one for
```
Updated comment to:
    // For local-dynamic on AIX, we need to generate one TOC entry for each
    // variable offset, and single module-handle TOC entry for the entire file.

orcguru commented:
> > The primary advantage of the local-dynamic access method is that you only make a single function call to __tls_get_mod() and use the returned pointer to get at all TLS variables. This implementation makes a function call for every variable. The only thing you've gained is a reduction in the number of TOC symbols.
> 
> The issue with duplicated calls to .__tls_get_mod is due to those pseudo TLSLDAIX nodes depending on different GVs, and CSE is not able to do dedup. I probably have to create a pseudo GV named like "_$TLSML", and use that GV for all TLSLDAIX pseudo nodes.

I tired TLSLDAIX without parameter (load module-handle is well known so the parameter is redundant), and now looks like those redundant .__tls_get_mod calls can be eliminated by optimization pass machine-cse.

orcguru commented:
I'm working on environment issue, and will update remaining two LIT cases and do testing etc.

I'm requesting another round of review in the mean time. Thank you all for your inputs!

orcguru commented:
> For below case:
> 
> ```
> __thread extern  int x = 42;
> __thread extern int y = 42;
> int main(void) {
>    return x + y;
> }
> ```
> 
> For now, we get:
> 
> ```
> .main:
> # %bb.0:                                # %entry
>         mflr 0
>         stwu 1, -32(1)
>         stw 0, 40(1)
>         lwz 6, L..C0(2)                         # target-flags(ppc-tlsld) @x
>         lwz 7, L..C1(2)                         # target-flags(ppc-tlsld) @y
>         lwz 3, L..C2(2)                         # target-flags(ppc-tlsld) &"_$TLSML[TC]"
>         bla .__tls_get_mod[PR]
>         add 3, 3, 6
>         lwz 6, 0(3)
>         lwz 3, L..C2(2)                         # target-flags(ppc-tlsld) &"_$TLSML[TC]"
>         bla .__tls_get_mod[PR]
>         add 3, 3, 7
>         lwz 3, 0(3)
>         add 3, 3, 6
>         addi 1, 1, 32
>         lwz 0, 8(1)
>         mtlr 0
>         blr
> ```
> 
> I think we should be able to save one `__tls_get_mod()`?
> 
> If this is not easy to optimize now, I'm OK we do this optimization later.

I think this has been addressed in [c5779c1](https://github.com/llvm/llvm-project/pull/66316/commits/c5779c1142936db45c1d0e2046ee5015a7d02fe9). TLSLDAIX nodes no longer have any parameter, they look the same to machine-cse, and so should be able to be eliminated now.

orcguru commented:
Currently obj mode has one issue: define more than one __thread variable, and all access return the first variable.

I'm investigating the root cause and will update patch accordingly. ASM mode is fine.

orcguru commented:
When a module contains three TLS LD variables

From assembly file generated by "as -a64 -many -o"
```
00000000000002f8 (idx: 33) a[TC]:
     2f8: 00 00 00 00   <unknown>
                00000000000002f8:  R_TLS_LD     (idx: 41) a[TL]
     2fc: 00 00 00 00   <unknown>

0000000000000300 (idx: 35) _$TLSML[TC]:
     300: 00 00 00 00   <unknown>
                0000000000000300:  R_TLSML      (idx: 35) _$TLSML[TC]
     304: 00 00 00 00   <unknown>

0000000000000308 (idx: 37) b[TC]:
     308: 00 00 00 00   <unknown>
                0000000000000308:  R_TLS_LD     (idx: 43) b[TL]
     30c: 00 00 00 04   <unknown>

0000000000000310 (idx: 39) c[TC]:
     310: 00 00 00 00   <unknown>
                0000000000000310:  R_TLS_LD     (idx: 45) c[TL]
     314: 00 00 00 08   <unknown>
```

Object mode output:
```
00000000000002f8 (idx: 31) a[TC]:
     2f8: 00 00 00 00   <unknown>
                00000000000002f8:  R_TLS_LD     (idx: 39) a[TL]
     2fc: 00 00 00 00   <unknown>

0000000000000300 (idx: 33) _$TLSML[TC]:
     300: 00 00 00 00   <unknown>
                0000000000000300:  R_TLSML      (idx: 33) _$TLSML[TC]
     304: 00 00 00 00   <unknown>

0000000000000308 (idx: 35) b[TC]:
     308: 00 00 00 00   <unknown>
                0000000000000308:  R_TLS_LD     (idx: 41) b[TL]
     30c: 00 00 00 00   <unknown>

0000000000000310 (idx: 37) c[TC]:
     310: 00 00 00 00   <unknown>
                0000000000000310:  R_TLS_LD     (idx: 43) c[TL]
     314: 00 00 00 00   <unknown>
```

Obj mode had wrong offset.

orcguru commented:
Missing an update in `recordRelocation()` similar as the patch: https://reviews.llvm.org/D155415.

Working on it now.

Commit:
Fix obj mode var access.

orcguru commented:
> Missing an update in `recordRelocation()` similar as the patch: https://reviews.llvm.org/D155415.
> 
> Working on it now.

Done. Added one test case aix-tls-ld-xcoff-reloc-large.ll to track this kind of issue. If full set of test case is necessary, please let me know, and I will add the remaining ones (reloc-large32, reloc, reloc-32).

orcguru commented on llvm/lib/Target/PowerPC/MCTargetDesc/PPCXCOFFObjectWriter.cpp:
```diff
@@ -116,6 +116,10 @@ std::pair<uint8_t, uint8_t> PPCXCOFFObjectWriter::getRelocTypeAndSignSize(
       return {XCOFF::RelocationType::R_TLS_IE, SignAndSizeForFKData};
     case MCSymbolRefExpr::VK_PPC_AIX_TLSLE:
       return {XCOFF::RelocationType::R_TLS_LE, SignAndSizeForFKData};
+    case MCSymbolRefExpr::VK_PPC_AIX_TLSLD:
```
I think there could be concern regarding this setting, and I observed obj output and the asm-as output is a little bit different on "IsSigned" setting on relocations for symbol ".__tls_get_mod", for example:

obj mode output processed by "llvm-readobj --relocs --expand-relocs"
```
  Section (index: 1) .text {
    Relocation {
      Virtual Address: 0xE
      Symbol: _$TLSML (11)
      IsSigned: No
      FixupBitValue: 0
      Length: 16
      Type: R_TOC (0x3)
    }
    Relocation {
      Virtual Address: 0x10
      Symbol: .__tls_get_mod (1)
      IsSigned: No
      FixupBitValue: 0
      Length: 26
      Type: R_RBA (0x18)
    }
    Relocation {
      Virtual Address: 0x16
      Symbol: a (13)
      IsSigned: No
      FixupBitValue: 0
      Length: 16
      Type: R_TOC (0x3)
    }
...
```

asm mode output assembled by "as -a64 -many ", and then processed by "llvm-readobj --relocs --expand-relocs"
```
  Section (index: 1) .text {
    Relocation {
      Virtual Address: 0x16
      Symbol: a (13)
      IsSigned: No
      FixupBitValue: 0
      Length: 16
      Type: R_TOC (0x3)
    }
    Relocation {
      Virtual Address: 0x1A
      Symbol: _$TLSML (15)
      IsSigned: No
      FixupBitValue: 0
      Length: 16
      Type: R_TOC (0x3)
    }
    Relocation {
      Virtual Address: 0x1C
      Symbol: .__tls_get_mod (3)
      IsSigned: Yes
      FixupBitValue: 0
      Length: 26
      Type: R_RBA (0x18)
    }
...
```

Notice they have different setting regarding "IsSigned" on the relocation for the ".__tls_get_mod" symbol.

I took another look into the behavior of general-dynamic, and then I saw the same difference there:

obj mode
```
    Relocation {
      Virtual Address: 0x1C
      Symbol: .__tls_get_addr (1)
      IsSigned: No
      FixupBitValue: 0
      Length: 26
      Type: R_RBA (0x18)
    }
```

asm mode
```
    Relocation {
      Virtual Address: 0x1C
      Symbol: .__tls_get_addr (3)
      IsSigned: Yes
      FixupBitValue: 0
      Length: 26
      Type: R_RBA (0x18)
    }
```

Looks like LD is aligned with GD in this particular behavior, so this may not be an issue.

orcguru replied:
Exactly! I was confused on this point back on [on Sep 26, 2023](https://github.com/llvm/llvm-project/pull/66316#discussion_r1336535324).

Definitely we need some setting here to avoid the message "Unsupported modifier". I'm lazy so I simply follow R_TLS_IE, and R_TLS_LE here.

Shall I set this to something other than SignAndSizeForFKData?

orcguru commented:
I saw comments being folded, paste here for notice, sorry about the duplication:

So I did have another version which can essentially remove the assert check with some side effects:

(1) It did generated extra external symbol reference, which looks like no harm with simple HelloWorld. (Notice: xlC does not generate this symbol reference.)
```
        .extern _Renamed..5f24__TLSML[UA]
        .rename _Renamed..5f24__TLSML[UA],"_$TLSML"
```
(2) Inside `XCOFFObjectWriter::recordRelocation()` logic I have to apply a hack to get the correct symbol, because now there are two symbols: one is `_Renamed..5f24__TLSML[UA]`, the other is `_Renamed..5f24__TLSML[TC]`. Without hack it will choose the [UA] one which will cause linker failure.

Attached the draft patch: [ReplaceAssertHack.diff.txt](https://github.com/llvm/llvm-project/files/12723368/ReplaceAssertHack.diff.txt)

I think current version has the advantage that it does not generate any extra external symbol reference or any extra relocation entry, although it looks ugly with regard to those asserts.

Is it possible that we fix the hack implementation with some future version (I will open an issue and work on it afterwards)? How about let's move on with current approach?

@amy-kwan @stephenpeckham @bzEq @chenzheng1030 Appreciate your comments. Thank you!

chenzheng1030 commented:
> Is it possible that we fix the hack implementation with some future version (I will open an issue and work on it afterwards)? How about let's move on with current approach?
> @amy-kwan @stephenpeckham @bzEq @chenzheng1030 Appreciate your comments. Thank you!

I am not against this proposal given the internal timeline.

chenzheng1030 commented:
The code formatting check fails.

orcguru commented:
> The code formatting check fails.

That is intentional: https://github.com/llvm/llvm-project/pull/66316#discussion_r1333195506

amy-kwan commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +160,92 @@ namespace {
+                int MatchCount = 0;
+                for (MachineOperand &MO : UseIter->operands()) {
+                  if (MO.isReg() && MO.isUse()) {
+                    if (RegInfo.hasOneDef(MO.getReg())) {
```
Pull out `MO.getReg()` calls into a separate variable after the `isReg()` check.

orcguru replied:
Updated. Thanks!

amy-kwan commented:
Also, is it intentional that we combined the patch to enable local-dynamic in clang within this patch?
Just wondering since I thought I saw a separate patch for the clang portion before (unless I am mistaken).

bzEq commented:
> Also, is it intentional that we combined the patch to enable local-dynamic in clang within this patch?

Yes. For github's nature, we are unable to set parent PR in llvm-project repo, so clang part is included in this PR.

orcguru commented on clang/test/Sema/aix-attr-tls_model.c:
```diff
@@ -6,6 +6,6 @@
 
 static __thread int y __attribute((tls_model("global-dynamic"))); // no-warning
-static __thread int y __attribute((tls_model("local-dynamic"))); // expected-error {{TLS model 'local-dynamic' is not yet supported on AIX}}
+static __thread int y __attribute((tls_model("local-dynamic"))); // expected-no-diagnostics
```
Thank you for asking. I did try to change it to "no-warning", however I got error message from llvm-lit complaining that since there is no diagnostic message anymore, we should "expected-no-diagnostics". So I did this change to make llvm-lit happy. If there is other formal/better setting, I will be happy to make the change.

orcguru replied:
The particular error for setting the check to "no-warning" is:

clang -cc1 -internal-isystem $llvm-public/build/lib/clang/18/include -nostdsysteminc -triple powerpc64-unknown-aix -target-cpu pwr8 -verify -fsyntax-only $llvm-project/clang/test/Sema/aix-attr-tls_model.c
error: no expected directives found: consider use of 'expected-no-diagnostics'
1 error generated.

orcguru commented:
> Also, is it intentional that we combined the patch to enable local-dynamic in clang within this patch? Just wondering since I thought I saw a separate patch for the clang portion before (unless I am mistaken).

Yes, I have to combine all changes into a monolithic patch, similar reason as mentioned by Kai. Otherwise those smaller patches will have to be serialized, review and commit one-by-one.

Commit:
[NFC] Incorporate comments.

orcguru commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +160,92 @@ namespace {
-          // The variable offset and region handle are copied in r4 and r3. The
-          // copies are followed by GETtlsADDR32AIX/GETtlsADDR64AIX.
-          if (!IsTLSTPRelMI) {
+          if (MI.getOpcode() == PPC::TLSLDAIX8 ||
```
Thanks! Updated.

orcguru commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +160,92 @@ namespace {
+            // It is better to put TLSLDAIX node before LoadOffsetToc node,
+            // because LoadOffsetToc node can use clobbers r4/r5. Search for the
+            // first paired LoadOffsetToc node within the same BB.
+            const PPCSubtarget &Subtarget =
```
Updated. Thanks!

Commit:
[NFC] Add FIXME to highlight existing issue:
The "_$TLSML" symbol did not lower through getTOCEntry().

amy-kwan commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -757,6 +772,17 @@ getTOCEntryTypeForMO(const MachineOperand &MO) {
   }
 }
+
+// FIXME: find alternative approach to get rid of this hack.
+// On AIX, TLS-local-dynamic requires that the symbol for the module handle must
+// have the name "_$TLSML". This symbol is used as one TOC symbol reference
+// itself with an ML relocation type, thus it has "[TC]" attached to its name.
```
```suggestion
// FIXME: find an alternative approach to get rid of this hack.
// On AIX, TLS-local-dynamic requires that the symbol for the module handle must
// have the name "_$TLSML". This symbol is used as one TOC symbol reference
// with an @ml relocation specifier, thus it has "[TC]" attached to its name.
```
Maybe this makes a bit more sense?

hubert-reinterpretcast commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -660,14 +671,16 @@ void PPCAsmPrinter::EmitTlsCall(const MachineInstr *MI,
+            (MI->getOperand(2).isReg() &&
+             MI->getOperand(2).getReg() == VarOffsetReg)) &&
            "GETtls[ld]ADDR[32] must read GPR4");
     EmitAIXTlsCallHelper(MI);
```
The helper functions have special calling convention properties. For example, they do not use the FP registers. The IBM XL compiler was able to take advantage of that.

For:
```c
__attribute__((tls_model("local-dynamic"))) __thread int x;
double g(int, double);
void f() {
  double gg = g(0, 1.);
  g(x, gg);
}
```

The IBM XL compilers were able to make use of the returned `double` staying in the register:
```
      28: 48 00 00 03   bla 0
                        00000028:  R_RBA        (idx: 36) .__tls_get_mod[PR]
      2c: 7c 66 18 2e   lwzx 3, 6, 3
      30: 4b ff ff d1   bl 0x0 <.f>
                        00000030:  R_RBR        (idx: 34) .g[PR]
```

Clang/LLVM loads the value from the stack:
```
      24: 48 00 00 03   bla 0
                0000000000000024:  R_RBA        (idx: 3) .__tls_get_mod[PR]
      28: e8 82 00 08   ld 4, 8(2)
                000000000000002a:  R_TOC        (idx: 17) x[TC]
      2c: 7c 63 22 aa   lwax 3, 3, 4
      30: 4b ff ff d1   bl 0x0 <.f>
                0000000000000030:  R_RBR        (idx: 1) .g[PR]
```

stephenpeckham replied:
From what I've observed, Clang/LLVM uses the special GPR usage rules for the call to __tls_get_mod.  But as you noted, FPRs are not handled as a special case.

orcguru replied:
Thank you for looking into this!

I tried the example, and it seems turn on optimization can help remove the FP load. We declared "Defs = [X0,X4,X5,X11,LR8,CR0]" for GETtlsMOD64AIX, and "Defs = [R0,R4,R5,R11,LR,CR0]" for GETtlsMOD32AIX. I think those FP registers should be treated as not touched by the call to __tls_get_mod.

hubert-reinterpretcast replied:
Thanks @orcguru. I misread the assembly 🤦‍♂️ (the extra instruction is not an `lfd`).

Commit:
Address the following FIXME:
The "_$TLSML" symbol did not lower through getTOCEntry().

The pseudo node "TLSLDAIX" now takes the _$TLSML GV node, and the
ppc-tls-dynamic-call pass is updated to fine tune the relative order
between the LoadOffset@toc node and the .__tls_get_mod node.

orcguru commented on llvm/lib/MC/XCOFFObjectWriter.cpp:
```diff
@@ -649,6 +649,14 @@ void XCOFFObjectWriter::recordRelocation(MCAssembler &Asm,
                                          uint64_t &FixedValue) {
   auto getIndex = [this](const MCSymbol *Sym,
                          const MCSectionXCOFF *ContainingCsect) {
+    // Fixup relocation flag for AIX TLS local-dynamic mode.
```
I will look into alternative approach to handle this fixup.

orcguru replied:
Proposed to fix the setting for the "_$TLSML" symbol in `TargetLoweringObjectFileXCOFF::getSectionForExternalReference()`.

MaskRay commented on llvm/lib/MC/XCOFFObjectWriter.cpp:
```diff
@@ -649,6 +649,14 @@ void XCOFFObjectWriter::recordRelocation(MCAssembler &Asm,
   auto getIndex = [this](const MCSymbol *Sym,
                          const MCSectionXCOFF *ContainingCsect) {
+    // Fixup relocation flag for AIX TLS local-dynamic mode.
+    if (Sym->getName().equals("_Renamed..5f24__TLSML[UA]")) {
```
We generally prefer `==` over `equals`

orcguru replied:
> We generally prefer `==` over `equals`

Thank you for pointing out!

MaskRay commented on llvm/lib/MC/XCOFFObjectWriter.cpp:
```diff
@@ -1360,6 +1372,9 @@ void XCOFFObjectWriter::assignAddressesAndIndices(const MCAsmLayout &Layout) {
   // Calculate indices for undefined symbols.
   for (auto &Csect : UndefinedCsects) {
+    // AIX does not need to emit for the _$TLSML symbol.
+    if (Csect.getSymbolTableName().equals(StringRef("_$TLSML")))
```
`== "..."`

MaskRay commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +160,78 @@ namespace {
+                       I != IE; ++I)
+                    if (LoadFromTocs.count(&*I)) {
+                      if (I->getOperand(1).isGlobal() &&
+                          I->getOperand(1).getGlobal()->getName().equals(
```
`== ...`

Commit:
Address comment: use == to compare StringRef names

Commit:
Correct SMC setting for the _$TLSML symbol

Commit:
Simplify and add check

Commit:
misc

Commit:
Simplify logic by move the XMC_TC setting for the _$TLSML symbol into getSectionForTOCEntry()

Commit:
Add check to make sure GV's name is defined before access

Commit:
update

orcguru commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -846,6 +859,19 @@ void PPCAsmPrinter::emitInstruction(const MachineInstr *MI) {
     if (MO.getTargetFlags() & PPCII::MO_TLSGD_FLAG)
       return MCSymbolRefExpr::VariantKind::VK_PPC_AIX_TLSGD;
+    if (MO.getTargetFlags() & PPCII::MO_TLSLD_FLAG) {
+      if (IsAIX && MO.isGlobal() && MO.getGlobal()->hasName() &&
```
This will be addressed after https://github.com/llvm/llvm-project/pull/69695

amy-kwan commented on llvm/lib/Target/PowerPC/MCTargetDesc/PPCMCTargetDesc.cpp:
```diff
@@ -231,12 +231,19 @@ class PPCTargetAsmStreamer : public PPCTargetStreamer {
+      // On AIX, we have TLS variable offsets (symbol@({gd|ie|le|ld}) depending
+      // on the TLS access method (or model). For the general-dynamic access
+      // method, we also have region handle (symbol@m) for each variable. For
+      // local-dynamic, there is a module handle (__TLSML[TC]@ml) for all
```
```suggestion
      // local-dynamic, there is a module handle (_TLSML[TC]@ml) for all
```
Single `_` I think?

amy-kwan commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -1356,6 +1382,11 @@ void PPCAsmPrinter::emitInstruction(const MachineInstr *MI) {
+  case PPC::GETtlsMOD64AIX:
+    // Transform: %r3 = GETtlsMODNNAIX %r3 (for NN == 32/64).
+    // Into: BLA .__tls_get_mod()
+    // Input parameter is a module handle (__TLSML[TC]@ml) for all variables.
```
```suggestion
    // Input parameter is a module handle (_TLSML[TC]@ml) for all variables.
```
Single `_`?

amy-kwan commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -2058,6 +2089,10 @@ void PPCAIXAsmPrinter::emitLinkage(const GlobalValue *GV,
   }
 
+  // Do not emit the _$TLSML symbol.
+  if (GVSym->getName() == "_Renamed..5f24__TLSML[TC]")
```
Maybe a silly question, but will this always contain the `5f24`?

orcguru replied:
I think `MCContext::createXCOFFSymbolImpl` is the logic that replaced the invalid characters. According to ASCII table, `_$` maps to `5f24`.

orcguru replied:
I think `MCContext::createXCOFFSymbolImpl` is the logic that replaced the invalid characters. According to ASCII table, `_$` maps to `5f24`...

stephenpeckham replied:
This check (and other checks in this PR) does a string comparison, which doesn't seem efficient.  Is there a way to set a flag on the symbol to avoid the string operations? Or are you out of flag bits?

orcguru replied:
Thank you! I will put some flag check before the string compare.

amy-kwan commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +160,78 @@ namespace {
-          if (!IsTLSTPRelMI) {
+          if (IsTLSLDAIXMI) {
+            // The relative order between the LoadOffset@toc node, and the
+            // ._tls_get_mod node is being tuned here. It is better to put the
```
```suggestion
            // .__tls_get_mod node is being tuned here. It is better to put the
```

amy-kwan commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +160,78 @@ namespace {
+            // LoadOffset@toc node after the call, since the LoadOffset@toc node
+            // can use clobbers r4/r5. Search for the pattern of two Load@toc
+            // nodes, and then move the LoadOffset@toc node right before the
+            // node that uses the OutReg of the ._tls_get_mod node.
```
```suggestion
            // node that uses the OutReg of the .__tls_get_mod node.
```

amy-kwan commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +160,78 @@ namespace {
+
+                // Check the two Load@toc: one should be _$TLSML, and the other
+                // will be moved before the node that uses the OutReg of the
+                // ._tls_get_mod node.
```
```suggestion
                // .__tls_get_mod node.
```

amy-kwan commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +160,78 @@ namespace {
+                                                   IE = MBB.end();
+                       I != IE; ++I)
+                    if (LoadFromTocs.count(&*I)) {
+                      if (I->getOperand(1).isGlobal() &&
```
Pull out `I->getOperand(1)` maybe?

amy-kwan reviewed (commented):
I imagine this will be updated after 4b932d84f48e0f3f42c769a5ca7ce6623ab62f2e so I plan to review again after.

Commit:
Apply target flag change && update according to comments

stephenpeckham commented on llvm/lib/CodeGen/TargetLoweringObjectFileImpl.cpp:
```diff
@@ -2645,10 +2653,14 @@ MCSection *TargetLoweringObjectFileXCOFF::getSectionForTOCEntry(
     const MCSymbol *Sym, const TargetMachine &TM) const {
   // Use TE storage-mapping class when large code model is enabled so that
   // the chance of needing -bbigtoc is decreased.
+  // The "_$TLSML" symbol for TLS local-dynamic mode requires XMC_TC.
```
You can use addis/load to load _$TLSML.  In this case, you can use XMC_TE.

orcguru replied:
When I change from XMC_TC to XMC_TE in the large code model like following,
```
L..C0:  
        .tc _Renamed..5f24__TLSML[TE],_Renamed..5f24__TLSML[TE]@ml
        .rename _Renamed..5f24__TLSML[TE],"_$TLSML"
```
The assembler complains like following:
```
Assembler:
tls_ld_te.s: line 62: 1252-215 The ML relocation type is only valid for a reference
        to a TOC symbol from the symbol itself.
```
Change back to "TC" did not see the message.

orcguru replied:
Looks like I have to keep the XMC_TC setting to make the assembler happy?

stephenpeckham replied:
You're right.  I didn't know the assembler would complain.

Commit:
Address comment: Add flag check before string compare.

Commit:
Update comment.

Commit:
[NFC] Add comments

Commit:
Remove dummy logic:
Since `getSectionForExternalReference` explicitly set XCOFF::XTY_SD for
the `_$TLSML` symbol reference, and that only CSType `XCOFF::XTY_ER`
symbol references are added into `UndefinedCsects`, the checks inside
`executePostLayoutBinding` are redundant.

Update test case to make sure external reference to the `_$TLSML` symbol
is not observed.

orcguru commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +164,89 @@ namespace {
-          // copies are followed by GETtlsADDR32AIX/GETtlsADDR64AIX.
-          if (!IsTLSTPRelMI) {
+          if (IsTLSLDAIXMI) {
+            // The relative order between the LoadOffset@toc node, and the
```
This is correct. I will elaborate in the comment.

orcguru commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +164,89 @@ namespace {
+              // Collect all instructions that use the OutReg.
+              for (MachineOperand &MO : RegInfo.use_operands(OutReg))
+                Uses.insert(MO.getParent());
+              // Find the first user (e.g.: lwax/stfdx) within the current BB.
```
Thank you! I will apply the change.

orcguru commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +164,89 @@ namespace {
+                if (Uses.count(&*UseIter))
+                  break;
+
+              // Got some work to do when UseIter pointing to valid node. Check
+              // the pattern and do the movement if match.
```
Yep. Much better! Thank you!

Commit:
[NFC] Update comments.

amy-kwan commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +164,94 @@ namespace {
+            // The call to .__tls_get_mod.
+            BuildMI(MBB, I, DL, TII->get(Opc2), GPR3).addReg(GPR3);
+          } else if (!IsTLSTPRelMI) {
+            // The variable offset and region handle are copied in r4 and r3.
```
nit: We use module handle in the other places but region handle here. I think we should use the same terminology.

orcguru replied:
Emm, according to the document (assembler-AIX-72.pdf), "region handle" and "region offset" are used to describe parameters for the __tls_get_addr() function.

orcguru replied:
Add comment "(for TLSGD)" to indicate this comment is not TLSLD related. Hope this is fine.

amy-kwan commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +164,94 @@ namespace {
+                if (LoadFromTocs.size() == 2) {
+                  MachineBasicBlock::iterator TLSMLIter = MBB.end();
+                  MachineBasicBlock::iterator OffsetIter = MBB.end();
+                  // Make sure the two LoadFromTocs are within current BB, and
+                  // one of them from the "_$TLSML" pseudo symbol, while the
+                  // other from the variable.
```
```suggestion
                  // Make sure the two LoadFromTocs nodes are within the current BB, and
                  // that one of them is from the "_$TLSML" pseudo symbol, while the
                  // other from the variable.
```

amy-kwan commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +164,94 @@ namespace {
+                      else
+                        OffsetIter = I;
+                    }
+                  // If both two iterators are valid, we should have identified
+                  // the scenario, and do the movement.
```
```suggestion
                  // Perform the movement when the desired scenario has been
                  // identified, which should be when both of the iterators are valid.
```
This may sound a bit better.

orcguru replied:
Yes, this is better indeed. Thank you!

amy-kwan reviewed (commented):
I believe some of the test cases also require updates as a result of https://github.com/llvm/llvm-project/pull/80162.

Commit:
rebase test case

orcguru commented:
> I believe some of the test cases also require updates as a result of #80162.

Sure. Updated!

Commit:
[NFC] Update comments.

Commit:
[NFC] Update comment.

amy-kwan commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +164,96 @@ namespace {
-          // copies are followed by GETtlsADDR32AIX/GETtlsADDR64AIX.
-          if (!IsTLSTPRelMI) {
+          if (IsTLSLDAIXMI) {
+            // The relative order between the LoadOffset@toc node (for the
+            // variable offset), and the .__tls_get_mod node is being tuned
+            // here. It is better to put the LoadOffset@toc node after the call,
+            // since the LoadOffset@toc node can use clobbers r4/r5. Search for
+            // the pattern of two Load@toc nodes (either for the variable offset
+            // or for the module handle), and then move the LoadOffset@toc node
+            // right before the node that uses the OutReg of the .__tls_get_mod
+            // node.
```
```suggestion
            // The relative order between the node that loads the variable offset
            // from the TOC, and the .__tls_get_mod node is being tuned here.
            // It is better to put the variable offset TOC load after the call,
            // since this node can use clobbers r4/r5.
            // Search for the pattern of the two nodes that load from the TOC
            // (either for the variable offset or for the module handle), and then
            // move the variable offset TOC load right before the node that uses the
            // OutReg of the .__tls_get_mod node.
```

amy-kwan commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +164,96 @@ namespace {
+              // of OutReg) is pointing to a valid node. Check the pattern and
+              // do the movement if the pattern matches.
+              if (UseIter != MBB.end()) {
+                // Collect associated Load@toc nodes. Use hasOneDef() to guard
```
```suggestion
                // Collect all associated nodes that load from the TOC. Use hasOneDef() to guard
```

amy-kwan commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +164,96 @@ namespace {
+                    if (RegInfo.hasOneDef(MOReg)) {
+                      MachineInstr *Temp =
+                          RegInfo.getOneDef(MOReg)->getParent();
+                      // For the current TLSLDAIX node, get the Load@toc node
+                      // for the InReg. Otherwise, Temp probably pointed to the
+                      // LoadOffset@toc node that we would like to move.
```
```suggestion
                      // For the current TLSLDAIX node, get the corresponding node that
                      // loads from the TOC for the InReg.
                      // Otherwise, Temp probably pointed to the variable offset TOC load
                      // that we would like to move.
```

amy-kwan commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +164,96 @@ namespace {
+                    }
+                  }
+
+                // Check the two Load@toc nodes: one should be _$TLSML, and the
+                // other will be moved before the node that uses the OutReg of
+                // the .__tls_get_mod node.
```
```suggestion
                // Check the two nodes that loaded from the TOC: one should be _$TLSML,
                // and the other will be moved before the node that uses the OutReg of
                // the .__tls_get_mod node.
```

amy-kwan commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +164,96 @@ namespace {
+                if (LoadFromTocs.size() == 2) {
+                  MachineBasicBlock::iterator TLSMLIter = MBB.end();
+                  MachineBasicBlock::iterator OffsetIter = MBB.end();
+                  // Make sure the two LoadFromToc nodes are within the current
+                  // BB, and that one of them is from the "_$TLSML" pseudo
+                  // symbol, while the other is from the variable.
```
```suggestion
                  // Make sure the two nodes that loaded from the TOC are within the current
                  // BB, and that one of them is from the "_$TLSML" pseudo
                  // symbol, while the other is from the variable.
```

amy-kwan reviewed (commented):
A question: For `PPCTLSDynamicCall.cpp`, would it be better if we instead reworded some of the sentences and not use `Load*@toc`?

I thought about this for a bit and I think the full sentences may be a bit better, so I added some sample comment suggestions, but if you think keeping the `Load*@toc` in the comments is better and more clear, then I think you can feel free to disregard the suggestion. However, if we are keeping the `Load*@toc`, it might be better to capitalize `TOC` instead.

orcguru commented on llvm/lib/Target/PowerPC/PPCISelLowering.h:
```diff
@@ -370,11 +370,22 @@ namespace llvm {
+    GET_TLS_MOD_AIX,
+
+    /// [GP|G8]RC = TLSLD_AIX, TOC_ENTRY(module handle)
+    /// Op that with single input of module handle TOC entry in R3, and
```
Sure. This is better. Thank you!

Commit:
[NFC] update comments

orcguru commented:
> A question: For `PPCTLSDynamicCall.cpp`, would it be better if we instead reworded some of the sentences and not use `Load*@toc`?
> 
> I thought about this for a bit and I think the full sentences may be a bit better, so I added some sample comment suggestions, but if you think keeping the `Load*@toc` in the comments is better and more clear, then I think you can feel free to disregard the suggestion. However, if we are keeping the `Load*@toc`, it might be better to capitalize `TOC` instead.

Updated according to the suggestions. Thank you!

amy-kwan commented on llvm/lib/Target/PowerPC/PPCTLSDynamicCall.cpp:
```diff
@@ -145,9 +164,97 @@ namespace {
+                    }
+                  }
+
+                // Check the two nodes that load from the TOC: one should be
+                // _$TLSML, and the other will be moved before the node that
```
```suggestion
                // Check the two nodes that loaded from the TOC: one should be
                // "_$TLSML", and the other will be moved before the node that
```
Minor comment update.

amy-kwan reviewed (approved):
Thank you for the updates!
For me personally, I have no further comments. Unless if anyone else has any comments, I think LGTM.

Commit:
[NFC] update comments

orcguru commented:
Thank you all for your efforts reviewing the patch!
If no objections, I will land it before the weekend. Thank you!

efriedma-quic commented:
It looks like a few of the regression tests are failing on the reverse-iteration buildbot (https://lab.llvm.org/buildbot/#/builders/54/builds/9683)

orcguru commented:
> It looks like a few of the regression tests are failing on the reverse-iteration buildbot (https://lab.llvm.org/buildbot/#/builders/54/builds/9683)

Thank you! Posted: https://github.com/llvm/llvm-project/pull/89714
